### PR TITLE
Stop sorting via `DefId`s in region resolution

### PR DESCRIPTION
**Pull request by oli-obk**
hopefully maintains the perf improvement from https://github.com/rust-lang/rust/pull/118824

works towards https://github.com/rust-lang/rust/issues/90317

### CODE REVIEW

File: compiler/rustc_infer/src/infer/lexical_region_resolve/mod.rs [hunk 1]
```diff
@@ -139,8 +139,8 @@ impl<'cx, 'tcx> LexicalResolver<'cx, 'tcx> {
         let mut var_data = self.construct_var_data();
 
         // Deduplicating constraints is shown to have a positive perf impact.
-        self.data.constraints.sort_by_key(|(constraint, _)| *constraint);
```
Comment (michaelwoerister) [inline compiler/rustc_infer/src/infer/lexical_region_resolve/mod.rs:142]: Would you mind using an [UnordSet](https://doc.rust-lang.org/stable/nightly-rustc/rustc_data_structures/unord/struct.UnordSet.html) here? It is a drop-in replacement in this case but guards against accidentally relying on its iteration order.

Comment (oli-obk) [reply compiler/rustc_infer/src/infer/lexical_region_resolve/mod.rs:142]: doesn't that have the same insertion issues as https://github.com/rust-lang/rust/pull/118824#issuecomment-1857798464 (expensive to insert)

Comment (oli-obk) [reply compiler/rustc_infer/src/infer/lexical_region_resolve/mod.rs:142]: or do you mean for the dedup set? sure, that doesn't matter

Comment (michaelwoerister) [reply compiler/rustc_infer/src/infer/lexical_region_resolve/mod.rs:142]: Yeah, I meant the `seen` set. The comment ended up in an ambiguous place 🙂

**Comment by michaelwoerister** [comment]
r=me if performance isn't impacted too much.

**Review by estebank — APPROVED** [review]
r=me

**Pull request merged**